internal/sci/aws/server.go

```diff
@@ -0,0 +1,218 @@
+// Package sciaws provides an AWS implementation of the Substratus Cloud Interface (SCI)
+package aws
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	awsSdk "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/eks"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/substratusai/substratus/internal/sci"
+)
+
+type Server struct {
+	sci.UnimplementedControllerServer
+	OIDCProviderURL string
+	OIDCProviderARN string
+	Clients
+}
+
+type Clients struct {
+	S3Client  *s3.S3
+	IamClient *iam.IAM
+}
+
+func (s *Server) GetObjectMd5(ctx context.Context, req *sci.GetObjectMd5Request) (*sci.GetObjectMd5Response, error) {
+	// ensure the object is accessible
+	bucketName, objectName := req.GetBucketName(), req.GetObjectName()
+	input := &s3.HeadObjectInput{
+		Bucket: awsSdk.String(bucketName),
+		Key:    awsSdk.String(objectName),
+	}
+	headResult, err := s.Clients.S3Client.HeadObject(input)
+	if err != nil {
+		return nil, err
+	}
+
+	// NOTE: AWS returns an MD5 checksum as an ETag except for multi-part uploads where it's an MD5 with a dash suffix.
+	if headResult.ETag == nil {
+		return nil, fmt.Errorf("object does not exist: %s", s3.ErrCodeNoSuchKey)
+	}
+
+	md5 := *headResult.ETag
+
+	return &sci.GetObjectMd5Response{
+		Md5Checksum: md5,
+	}, nil
+}
+
+func (s *Server) CreateSignedURL(ctx context.Context, req *sci.CreateSignedURLRequest) (*sci.CreateSignedURLResponse, error) {
+	bucketName, objectName, checksum := req.GetBucketName(),
+		req.GetObjectName(),
+		req.GetMd5Checksum()
+
+	// ensure the object is accessible
+	input := &s3.HeadObjectInput{
+		Bucket: awsSdk.String(bucketName),
+		Key:    awsSdk.String(objectName),
+	}
+	_, err := s.Clients.S3Client.HeadObject(input)
```

Comment:
What does this check tell us after we get past it?

Reply:
Like in the gcs sci, I added this to catch errors related to accessing this key path.

Some S3 errors we might catch:

Permission issues (maybe the provided credentials don't have the required permission). Network issues (maybe there's an issue connecting to the S3 service). Bucket issues (unlikely but potentially the bucket doesn't exist). AWS SDK or client-related issues.

Non-AWS Errors: Errors not related to the AWS SDK. Potentially lower-level issues

Reply:
I would prefer to catch permission issues in Validate function that only runs once on startup. That's what I ended up having to do for sci-gcp.

Reply:
Lets remove this check. We are adding network calls for every request and it seems counter-intuitive to me as a reader to run a check that looks up an object that will not exist during happy path. If the bucket does not exist, I would expect the following call to generate a signed url to fail. If there is a networking issue it will be log eventually, perhaps in the next call to create a signed url, perhaps if a call to get the md5. If the SCI is lacking permissions that too will surface when fetching the md5.